core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala

```diff
@@ -1928,9 +1928,15 @@ class TaskSetManagerSuite
     // get the TaskSetManager
     val manager = sched.invokePrivate(taskSetManagers()).get(stageId).get(stageAttemptId)
 
-    val task0 = manager.resourceOffer(exec0, "localhost", TaskLocality.NO_PREF)
-    val task1 = manager.resourceOffer(exec1, "localhost", TaskLocality.NO_PREF)
-    assert(task0.isDefined && task1.isDefined)
+    var task0 : Option[TaskInfo] = None
+    var task1 : Option[TaskInfo] = None
+
+    eventually(timeout(10.seconds), interval(100.milliseconds)) {
+      task0 = Option(manager.taskInfos(0))
+      task1 = Option(manager.taskInfos(1))
+      assert(task0.isDefined && task1.isDefined)
+    }
```

Comment:
The race condition is between `sched.submitTasks` and `manager.resourceOffer`. That means, tasks within taskSet can be scheduled or even finished before we call `manager.resourceOffer`. If this happens, then `manager.resourceOff` would return `None`. In a worse case, the taskSet can be finished before we call `manager.resourceOffer`, then, `manager` can be null when we call `sched.invokePrivate(taskSetManagers()).get(stageId).get(stageAttemptId)`.  So it's still possible with this fix that tasks has already finished when we assert running status below.

BTW, I believe that we don't need to wait `managers` to return `taskInfos`, as it always exists once we submit the taskset to `TaskSchedulerImpl`.

Reply:
Ok i see, i have used a Fake long running task now, which will not cause this issue you mentioned.